Verejne clickable

Siegrift opened the pull request:
You can finally click on the markers on the map and side panel

mpinter commented on client/src/components/Verejne/entityHelpers.js:
```diff
@@ -0,0 +1,16 @@
+// Entity which is a number is individual.
+// District starts with letter
+export const isIndividual = (eid) => {
+  return !isNaN(parseFloat(eid)) && isFinite(eid)
```
this seems like a weird check for what you're saying you need to check ... why the finite test ?

mpinter commented on client/src/components/Verejne/entityHelpers.js:
```diff
@@ -0,0 +1,16 @@
+}
+
+export const isPolitician = (entity) => {
+  return entity.ds != null && entity.ds.length >= 1 && entity.ds[0] === 1
```
please no, eslint should forbid you from using != and == either way

mpinter commented on client/src/components/Verejne/entityHelpers.js:
```diff
@@ -0,0 +1,16 @@
+}
+
+export const hasContractsWithState = (entity) => {
+  if (entity.ds != null && entity.ds.length >= 2 && isIndividual(entity.eid)) {
```
same here

mpinter commented on client/src/components/Verejne/entityHelpers.js:
```diff
@@ -0,0 +1,16 @@
+// Entity which is a number is individual.
```
I assume you did not touch this file, but if we're using it, please fix and add flow

mpinter commented on client/src/components/Verejne/index.js:
```diff
@@ -5,29 +5,54 @@ import './Verejne.css'
-const Verejne = ({entities, fetchEntities, pageCount, currentPage, setCurrentPage}) => (
+const renderListItemIcon = (entity) => {
+  if (entity.size > 1) {return <img src={MapIcon} style={{width: '2rem', height: '2rem'}} alt="listItemIcon" />}
+  const color = isPolitician(entity) ? '#e55600' : '#0062db'
```
maybe do with classes (we're using classnames so it'll be nice), and we can then have these colors in variables

mpinter commented on client/src/components/Verejne/index.js:
```diff
@@ -45,20 +70,25 @@ export default compose(
-    (props) => !props.entities.length,
-    renderComponent(Loading),
-  )
+  withProps(({entities, currentPage}) => {
```
I see you're already using redux - then please do this with a selector instead of withProps. It both makes more sense organization wise and has the additional benefit of memoization and composability

mpinter replied:
also, I'm kind of missing the flow annotations across the board in this file

mpinter commented on client/src/components/Verejne/index.js:
```diff
@@ -45,20 +70,25 @@ export default compose(
+      pageCount: entities ? Math.ceil(entities.length / VEREJNE_MAX_PAGE_ITEMS) : 0,
+    }
+  }),
+  branch((props) => !props.entities, renderComponent(Loading))
```
this is nice 👍 .. though it'll get handled by data-providers once they're here, good to know

mpinter commented on client/src/components/Verejne/Legend/index.js:
```diff
@@ -6,6 +6,7 @@ import classnames from 'classnames'
 import './Legend.css'
+import MapIcon from '../mapIcon.svg'
 
 const Legend = ({legendOpen, setLegendOpen}) => (
```
flow

mpinter commented on client/src/components/Verejne/GoogleMap/index.js:
```diff
@@ -43,6 +55,11 @@ const getClusters = (mapOptions: MapOptions, entities: Array<Entity>): Array<Clu
   return clusters(mapOptions)
 }
 
+const getClusterTooltip = (entity): string => {
```
flow for param

mpinter commented on client/src/components/Verejne/GoogleMap/index.js:
```diff
@@ -58,11 +75,62 @@ const createClusters = (mapOptions: MapOptions, entities: Array<Entity>): Array<
   return clusters
 }
 
+const getEntityMarker = (entity): string => {
```
also flow through the rest of this file

mpinter commented on client/src/components/Verejne/GoogleMap/Marker/Marker.scss:
```diff
@@ -25,14 +25,30 @@
+        margin: auto;
+    }
+    &__Normal svg {
+        color: $marker-color !important;
```
you are using quite a lot of importants in css

my assumption is there are some hard to override defaults on the google maps component, so it's probably the only way, but still - think over other options (adding another custom class?), because this can get hard to maintain

mpinter commented on client/src/components/Verejne/GoogleMap/index.js:
```diff
@@ -58,11 +75,62 @@ const createClusters = (mapOptions: MapOptions, entities: Array<Entity>): Array<
+          else zoomToLocation({lat: e.lat, lng: e.lng})
+        }}
+      >
+        {e.numPoints !== 1 && <span className="Marker__Text">{e.numPoints}</span>}
```
use ternary

mpinter commented on client/src/components/Verejne/GoogleMap/index.js:
```diff
@@ -43,6 +55,11 @@ const getClusters = (mapOptions: MapOptions, entities: Array<Entity>): Array<Clu
+  return sorted[0].name
+}
+
 const createClusters = (mapOptions: MapOptions, entities: Array<Entity>): Array<MapCluster> => {
```
I probably missed earlier (or ignored because redux wasn't in play yet) but this belongs to a selector

mpinter commented on client/src/components/Verejne/GoogleMap/index.js:
```diff
@@ -58,11 +75,62 @@ const createClusters = (mapOptions: MapOptions, entities: Array<Entity>): Array<
+        lat={e.lat}
+        lng={e.lng}
+        title={getClusterTooltip(e)}
+        onClick={() => {
```
passing arrow function to onClick of react-component may hinder performance

mpinter commented on client/src/components/Verejne/GoogleMap/index.js:
```diff
@@ -58,11 +75,62 @@ const createClusters = (mapOptions: MapOptions, entities: Array<Entity>): Array<
+
+const renderMarkers = (mapOptions, entities, selectEntity, zoomToLocation) => {
+  const zoom = mapOptions.zoom
+  const clusters = map(createClusters(mapOptions, entities))
```
doing this on every render definitely hinders performance

mpinter commented on client/src/components/shared/GeneralPagination/index.js:
```diff
@@ -37,29 +38,34 @@ const getNextPage = ({currentPage, page, from, to, pageCount}) => {
+      {range(from, to).map((page) => (
+        <PaginationItem
+          key={page}
+          onClick={() =>
```
again, was here in previous commit, but you shouldn't pass arrow functions to onClick of react components

that said, this is the pagination that will get a rewrite, so no problem for keeping it for now, just take note

Siegrift commented:
Sorry for a lot of messy stuff in the first process, I shouldn't have pushed so early... Fixed almost all your points:
- rewrote the legacy helpers (except one '==' which I want to check from guys what to expect)
- added flow everywhere it was missing, fixed long term flow errors (should have 0 :) )
- didn't remove !important but I reduced total occurrences a bit, if you know a neat way how to avoid them tell me
- split components
- optimized arrow functions in marker, but that's not the preformance hinder here. I am quite sure it's the rendering, although I get the point that hurt more in the future.

mpinter commented on client/src/components/Verejne/entityHelpers.js:
```diff
@@ -0,0 +1,16 @@
+
+// entity which is a number is individual, district starts with letter
+export const isIndividual = (eid: string): boolean => {
+  return '0123456789'.indexOf(eid[0]) === -1
```
heh, nice

mpinter commented on client/src/components/Verejne/entityHelpers.js:
```diff
@@ -0,0 +1,16 @@
+
+export const isPolitician = (entity: Entity): boolean => {
+  // TODO: == is on purpose as this field can be: null, 0, true...
+  return entity.ds && entity.ds[0] == 1
```
and what do you check for exactly ? does entity.ds === 0 mean that entity is a politician or not ? what is the case where !!entity.ds does not work ?

mpinter commented on client/src/components/Verejne/entityHelpers.js:
```diff
@@ -0,0 +1,16 @@
+}
+
+export const hasContractsWithState = (entity: Entity): boolean => {
+  return entity.ds && isIndividual(entity.eid) && entity.ds[1] > 0
```
are you sure ds has at least 2 chars ? and why checking > 0 ? (at least add a comment regarding the second question to the code)

mpinter commented on client/src/components/Verejne/index.js:
```diff
@@ -5,35 +5,68 @@ import './Verejne.css'
+  return <Icon size="18" className={classnames('SidePanel__List__Item__Icon', color)} />
+}
+
+// NOTE: there can be multiple points in the map on the same location...
```
*...points on the map... :)

mpinter commented on client/src/components/Verejne/index.js:
```diff
@@ -42,35 +75,28 @@ const Verejne = ({
+      mapReference: mapReferenceSelector(state),
     }),
-    {fetchEntities}
+    {selectEntity, setCurrentPage}
```
does mapDispatchToProps work this way ? (I'm kind of hoping this is some new syntax I don't know about where it takes dispatch from context, but probably just typo)

mpinter commented on client/src/components/Verejne/index.js:
```diff
@@ -42,35 +75,28 @@ const Verejne = ({
       />
     </div>
-    <GoogleMap />
+    <GoogleMap refetch={() => refetch(getEntitiesUrlFromMapReference(mapReference))} />
```
just note that again, you'd force a re-render every time Verejne rerenders because you're using and arrow function, but more on passign refetch elsewhere

mpinter commented on client/src/components/shared/GeneralPagination/index.js:
```diff
@@ -0,0 +1,73 @@
+// @flow
```
rebase and use the pagination lib from master, should be able to do all the stuff needed

mpinter commented on client/src/dataProviders/publiclyDataProviders.js:
```diff
@@ -0,0 +1,76 @@
+  usedLevel: number,
+}
+
+const getZoomLevel = (mapReference: MapReference): number => {
```
this looks like a selector

mpinter commented on client/src/dataProviders/publiclyDataProviders.js:
```diff
@@ -0,0 +1,76 @@
+  return [ENTITY_ZOOM, SUB_CITY_ZOOM, CITY_ZOOM].filter((val) => val > zoom).length
+}
+
+const getRequestParams = (mapReference: MapReference): EntitiesRequestParams => {
```
also a selector

mpinter commented on client/src/dataProviders/publiclyDataProviders.js:
```diff
@@ -0,0 +1,76 @@
+  return params
+}
+
+const dispatchEntities = () => (ref: string, data: Array<Entity>, dispatch: Dispatch) =>
```
I'd maybe rather dispatch setEntities directly and do the mapping there?

mpinter commented on client/src/dataProviders/publiclyDataProviders.js:
```diff
@@ -0,0 +1,76 @@
+      },
+    ],
+    onData: [dispatchEntities],
+    keepAliveFor: 2000,
```
not much reason to keepAliveFor 2 seconds, that's practically the same as 0

mpinter commented on client/src/selectors/index.js:
```diff
@@ -70,4 +71,66 @@ export const mapOptionsSelector = (state: State): MapOptions => state.mapOptions
+  (entities) => (entities ? Math.ceil(entities.length / VEREJNE_MAX_PAGE_ITEMS) : 0)
+)
+
+// TODO performance
```
an easy performance save is to make a selector for entities sorted by size, and then use this one just to chunk and paginate

Siegrift commented on client/src/components/Verejne/GoogleMap/ClusterMarker/index.js:
```diff
@@ -0,0 +1,73 @@
+  onClick: () => void,
+}
+
+const getClusterTooltip = (cluster: MapCluster): string => {
```
Mhh, how do you mean? This is something derived props.

mpinter commented on client/src/components/Verejne/Map/index.js:
```diff
@@ -0,0 +1,111 @@
+}
+
+export default compose(
+  connect(
```
please split this connects as written in previous comment (now outdated as the original file was split)
